File: src/node/magma/Lang.ts
/*import magma.rule.divide.divide.DivideRule;*/
/*import magma.rule.InfixRule;*/
/*import magma.rule.OrRule;*/
/*import magma.rule.PlaceholderRule;*/
/*import magma.rule.PrefixRule;*/
/*import magma.rule.Rule;*/
/*import magma.rule.StringRule;*/
/*import magma.rule.StripRule;*/
/*import magma.rule.SuffixRule;*/
/*import magma.rule.TypeRule;*/
/*import java.util.List;*/
/*public*/class Lang {/*static DivideRule createTSRootRule() {
		return new DivideRule("children", new SuffixRule(Lang.createTSRootSegmentValueRule(), System.lineSeparator()));
	}

	static DivideRule createJavaRootRule() {
		return new DivideRule("children", Lang.createJavaRootSegmentValueRule());
	}

	private static Rule createJavaRootSegmentValueRule() {
		return new OrRule(
				List.of(Lang.createJavaClassRule(), Lang.createPackageRule(), new StripRule(Lang.createPlaceholderRule())));
	}

	private static TypeRule createPackageRule() {
		return new TypeRule("package", new StripRule(new PrefixRule("package ", new StringRule("value"))));
	}

	private static Rule createTSRootSegmentValueRule() {
		return new OrRule(List.of(Lang.createTSClassRule(), Lang.createPlaceholderRule()));
	}

	private static Rule createPlaceholderRule() {
		return new TypeRule("placeholder", new PlaceholderRule(new StringRule("value")));
	}

	private static Rule createJavaClassRule() {
		final var modifiers1 = new StringRule("modifiers");
		final var name =
				new InfixRule(new StringRule("name"), "{", new StripRule(new SuffixRule(new StringRule("with-end"), "}")));
		return new TypeRule("class", new InfixRule(modifiers1, "class ", name));
	}

	private static Rule createTSClassRule() {
		final var content = new PlaceholderRule(new StringRule("with-end"));
		final var name = new InfixRule(new StringRule("name"), " {", new SuffixRule(content, "}"));
		final var modifiers = new PlaceholderRule(new StringRule("modifiers"));
		return new TypeRule("class", new InfixRule(modifiers, "class ", name));
	}
*/}

File: src/java/magma/Lang.java
package magma;

import magma.rule.InfixRule;
import magma.rule.OrRule;
import magma.rule.PrefixRule;
import magma.rule.Rule;
import magma.rule.StringRule;
import magma.rule.StripRule;
import magma.rule.SuffixRule;
import magma.rule.TypeRule;
import magma.rule.divide.NodeListRule;

import java.util.List;

/**
 * Utility class that provides factory methods for creating parsing rules
 * for Java and TypeScript code. This class contains methods for creating
 * various rule configurations that can be used to lex source code into
 * an abstract syntax tree (AST) and generate code from that AST.
 * <p>
 * The class handles rules for Java and TypeScript language constructs
 * such as packages, imports, classes, and placeholders.
 */
final class Lang {
	private Lang() {}

	static NodeListRule createJavaRootRule() {
		return new NodeListRule(Lang.createJavaRootSegmentRule(), "children");
	}

	private static OrRule createTSRootSegmentRule() {
		return new OrRule(List.of(Lang.createTSClassRule(), Lang.createTypedPlaceholderRule()));
	}

	private static OrRule createJavaRootSegmentRule() {
		return new OrRule(
				List.of(Lang.createNamespacedRule("package"), Lang.createNamespacedRule("import"), Lang.createJavaClassRule(),
								Lang.createTypedPlaceholderRule()));
	}

	private static TypeRule createTypedPlaceholderRule() {
		return new TypeRule("placeholder", new StringRule("content"));
	}

	private static Rule createNamespacedRule(final String type) {
		final var content = new PrefixRule(type + " ", new SuffixRule(new StringRule("content"), ";"));
		return new TypeRule(type, new StripRule(content));
	}

	private static Rule createTSClassRule() {
		final Rule name = new StringRule("name");
		final Rule body = new NodeListRule(Lang.createTSClassMemberRule(), "children");

		return new TypeRule("class", new InfixRule(new SuffixRule(new PrefixRule("export class ", name), " {"), "",
																							 new SuffixRule(body, "}")));
	}

	private static TypeRule createTSClassMemberRule() {
		return Lang.createTypedPlaceholderRule();
	}

	private static Rule createJavaClassRule() {
		final Rule name = new StringRule("name");
		final Rule body = new NodeListRule(Lang.createJavaClassMemberRule(), "children");

		// More flexible rule that can handle modifiers before "class"
		// and additional content between class name and opening brace
		return new InfixRule(new StringRule("modifiers"), "class ",
												 new InfixRule(name, "{", new StripRule(new SuffixRule(body, "}"))));
	}

	private static TypeRule createJavaClassMemberRule() {
		return Lang.createTypedPlaceholderRule();
	}

	static NodeListRule createTSRootRule() {
		return new NodeListRule(Lang.createTSRootSegmentRule(), "children");
	}
}
